### PR TITLE
Concurrency bug-squashing

### PR DESCRIPTION
**Pull request by grdw**
Fixes: #1145 
Fixes: #1170

### CODE REVIEW

**Comment by antw** [comment]
First impressions are that this is an improvement, but still not quite there yet...

1. I tried to recreate the issue described in https://github.com/quintel/etmoses/issues/1145#issuecomment-230720172:

  > What I found so far is that if I edit the LES and set solar PV concurrency to "min" and hit "Apply" everything seems to work fine. However, if I add a solar PV array to "Endpoint 2" and then change the concurrency setting, the heating technologies on "Endpoint 1" are split up. It seems half the heating technologies are assigned to a non-existent space_heating_buffer_2.

  The good news is that heating technologies no longer get split between the original buffer and a non-existent `buffer_space_heating_2`. However, the buffer does get renamed from `buffer_space_heating_1` to `buffer_space_heating_2`. The buffer number seems to increment by one each time the concurrency settings are toggled and applied.

2. To [the test scenario](https://beta-moses.energytransitionmodel.com/testing_grounds/585), add a single *empty* "Buffer space heating" to "Endpoint 2" and toggle Solar PV concurrency to "min"; heating technologies and buffers are still unchanged at the "max" setting. The heating technologies from the buffer on "Endpoint 1" are *duplicated* onto the buffer on "Endpoint 2" (i.e. if there were 5 units of "space heater combined network gas" on "Endpoint 1", there are now an additional 5 units also installed on "Endpoint 2".

3. Add a second "Buffer space heating" to "Endpoint 1", set demand to `1000` and hit "Apply" on the concurrency settings (no changes made). The original buffer is removed, and all the technologies are moved to the new one.

  I guess that's expected since the concurrency setting for that technology is "max", but for buffers I would expect the demand, volume (and perhaps capacity?) attributes to be summed (i.e. two buffers with demand of 1000 should be merged into one with 2000).

4. Similarly, without making any changes to the test scenario, if "Buffer space heating" is toggled to "min" concurrency, there are now five buffers all with the original amount of demand (total heat demand for the LES is now 5x higher than before).

*3 and 4 also happen on production/beta, so those are not new problems.*

--

This might not be a popular sentiment given the work which has gone into this, but I really think there's a strong case for removing the concurrency settings entirely:

* Using "min" concurrency on anything but the smallest of LESes causes performance problems with the technology form, and results in excessively long network calculations.

* As (3) demonstrates, it's very easy to lose data when switching to "max" concurrency.

  A simpler example: add 2 P2P batteries and set a custom capacity on one. Apply the concurrency settings. You either end up with two custom-capacity batteries, or two default-capacity batteries (depending on which order they appeared on the form).

I would be curious to know how these settings are used in the real-world by our visitors, because I think we could find simpler alternatives for splitting techs into smaller units.

(pinging @ChaelKruip and @dennisschoenmakers)

File: app/services/testing_ground/concurrency/maximizer.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class TestingGround
+  class Concurrency
+    module Maximizer
+      def maximize
+        -> (techs) {
+          tech       = techs.first
+          tech.units = techs.sum(&:units)
+          tech
+        }
```
Comment (antw) [inline app/services/testing_ground/concurrency/maximizer.rb]: Very minor thing: use `-> { }` for single-line blocks, and `lambda do ... end` for multi-line blocks.

https://github.com/bbatsov/ruby-style-guide#lambda-multi-line

**Comment by grdw** [comment]
> [..] there are now an additional 5 units also installed on "Endpoint 2".

I just found out point 2 as well .. Working on it as we speak.

**Comment by grdw** [comment]
> but I really think there's a strong case for removing the concurrency settings entirely:

I would have no problem with this personally, but I think that this might be a different problem. 

> LESes causes performance problems with the technology form

Yes, but this is also the case if you have a topology with > 100 end-points. I think we haven't even tried a 1000 end-point topology. The concurrency causes a large form, but it's not the only reason which can cause a large form to render.

> I would be curious to know how these settings are used in the real-world by our visitors, because I think we could find simpler alternatives for splitting techs into smaller units.

Hmm... I get what you are saying here but I wonder what hose simpler alternatives would be? Maybe if the user specifies manually in groups of how many units someone wants to split their technologies?

**Comment by ChaelKruip** [comment]
> This might not be a popular sentiment given the work which has gone into this, but I really think there's a strong case for removing the concurrency settings entirely:

I think we should consider the possibility. I'm not against it in any case.

**Comment by antw** [comment]
> I guess that's expected since the concurrency setting for that technology is "max", but for buffers I would expect the demand, volume (and perhaps capacity?) attributes to be summed (i.e. two buffers with demand of 1000 should be merged into one with 2000).

I don't know what I was thinking when I wrote this. If the two buffers have the same demand, setting `units` to 2 would have to correct effect of doubling the demand. Mind you, that doesn't help if the two buffers have different demands, but that's a another problem entirely.

> Hmm... I get what you are saying here but I wonder what hose simpler alternatives would be? Maybe if the user specifies manually in groups of how many units someone wants to split their technologies?

I think going from min to max concurrency is problematic, since visitors will almost certainly lose data (custom capacities, demands, etc). I would be in favour of a newly-imported scenario being as it is today: max concurrency. Then, if a user desires something more like "min" concurrency, they can hit a "Split" button next to the technology:

screen shot 2016-07-08 at 12 02 14

A technology with `units = 4` becomes two identical technologies with `units = 2`. Then the visitor can customise each one as they desire. If they want, they could then split those into four units of 1.

On the back-end when the LES is saved we would *ideally* merge *identical* technologies together. Two identical solar PV techs could be merged into one with `units = 2`.

**Comment by grdw** [comment]
> they can hit a "Split" button next to the technology:

I kind of like this (how it looks :D ) but what happens after someone has split a technology? How do you return back to normal? Or if you hit split twice in a row, nothing happens?

**Comment by grdw** [comment]
I'd suggest to maybe put this in another issue and keep the current concurrency things as they  are. And maybe later we could actually change it into something better? @ChaelKruip what do you say?

**Comment by ChaelKruip** [comment]
> I'd suggest to maybe put this in another issue and keep the current concurrency things as they are. And maybe later we could actually change it into something better? @ChaelKruip what do you say?

Yes. @grdw can you do this?

**Comment by antw** [comment]
I saw that changes have been pushed, so I tried the [same steps I reported above](https://github.com/quintel/etmoses/pull/1216#issuecomment-231099936):

1. Is fixed!

2. Kind of fixed, but problem (1) makes an appearance. "Buffer space heating 1" on the first endpoint has been renamed to "Buffer space heating 2", and the "Buffer space heating 2" on the second endpoint has been renamed to "Buffer space heating 3"

  Also: Set "Buffer space heating" concurrency to "min" and apply. There are now five buffers (which is correct) with the names "Buffer space heating 6", "Buffer space heating 7", ...

3. The newly-added buffer is renamed (from "2" to "3"), and the original buffer **and attached heating technologies** all disappear.

4. **New**: Set any one of the technologies attached to the buffer to have `units = 1`, "Buffer space heating" concurrency to "min" and apply. There is one instance of the technology attached to the first buffer with `units = 1`. The other four buffers also have an instance of the technology technology with `units = 0`; I don't think those technologies should exist.

If we're not going to turn off the concurrency feature, I still think we should push ahead with merging this. However I do think we should get rid of it ASAP because it's too easy for visitors to shoot themselves in the foot (and a pain for us to maintain and test).

**Comment by ChaelKruip** [comment]
> If we're not going to turn off the concurrency feature, I still think we should push ahead with merging this. 

Let's do this for now.

> However I do think we should get rid of it ASAP because it's too easy for visitors to shoot themselves in the foot (and a pain for us to maintain and test).

I agree but we should present the client with an alternative idea.

**Comment by antw** [comment]
> I agree but we should present the client with an alternative idea.

I'll document my proposal more thoroughly in a new issue later.

I do think (3) needs to be fixed before merging; we can't have entire heat buffers disappearing...

**Comment by grdw** [comment]
It's a little unclear to me what we're going to do with this PR? Can we just close it (without merging) and start with the simple 'split' idea of Anthony any time soon? Or should we try to fix point number 3, merge it and than replace it with the 'split' idea. 

It all comes down a little in which time frame we're going to do the actual replacement. If it's going to take at least 2-3 months before we start with that I'd say, fix (3) and merge it. If it's any time soon than I'd say close it.

CC: @ChaelKruip

**Comment by antw** [comment]
> It all comes down a little in which time frame we're going to do the actual replacement. If it's going to take at least 2-3 months before we start with that I'd say, fix (3) and merge it. If it's any time soon than I'd say close it.

The sprint plan includes "Concurrency-code should be replaced by something less error-prone", so hopefully we might get to it this month? I'll post an issue describing the "Split" button I proposed and we can discuss it.

If it is indeed going to be more than a month, I agree that we should merge this after fixing (3), because it still works much better than the old concurrency code.

**Comment by ChaelKruip** [comment]
> The sprint plan includes "Concurrency-code should be replaced by something less error-prone", so hopefully we might get to it this month? I'll post an issue describing the "Split" button I proposed and we can discuss it.

👍 

I think we should close this and start from scratch soon!

**Pull request closed without merging**